bindings/rust/evmc-declare/src/lib.rs

```diff
@@ -0,0 +1,383 @@
+            custom_version: vm_version_string_optional,
+        }
+    }
+
+    fn get_capabilities(&self) -> u32 {
+        self.capabilities
+    }
+
+    fn get_name_stylized(&self) -> &String {
+        &self.name_stylized
+    }
+
+    fn get_custom_version(&self) -> &Option<String> {
+        &self.custom_version
+    }
+}
+
+#[proc_macro_attribute]
+pub fn evmc_declare_vm(args: TokenStream, item: TokenStream) -> TokenStream {
+    // First, try to parse the input token stream into an AST node representing a struct
+    // declaration.
+    let input: ItemStruct = parse_macro_input!(item as ItemStruct);
+
+    // Extract the identifier of the struct from the AST node.
+    let vm_type_name: String = input.ident.to_string();
+
+    // Build the VM name set.
+    let names = VMNameSet::new(vm_type_name);
+
+    // Parse the arguments for the macro.
+    let meta_args = parse_macro_input!(args as AttributeArgs);
+    let vm_data = VMMetaData::new(meta_args);
+
+    // Get all the tokens from the respective helpers.
+    let static_data_tokens = build_static_data(&names, &vm_data);
+    let capabilities_tokens = build_capabilities_fn(vm_data.get_capabilities());
+    let create_tokens = build_create_fn(&names);
+    let destroy_tokens = build_destroy_fn(&names);
+    let execute_tokens = build_execute_fn(&names);
+
+    let quoted = quote! {
+        #input
+        #static_data_tokens
+        #capabilities_tokens
+        #create_tokens
+        #destroy_tokens
+        #execute_tokens
+    };
+
+    quoted.into()
+}
+
+/// Generate tokens for the static data associated with an EVMC VM.
+fn build_static_data(names: &VMNameSet, metadata: &VMMetaData) -> proc_macro2::TokenStream {
+    // Stitch together the VM name and the suffix _NAME
+    let static_name_ident = names.get_caps_as_ident_append("_NAME");
+    let static_version_ident = names.get_caps_as_ident_append("_VERSION");
+
+    // Turn the stylized VM name and version into string literals.
+    let stylized_name_literal = LitStr::new(
+        metadata.get_name_stylized().as_str(),
+        metadata.get_name_stylized().as_str().span(),
+    );
+
+    // If the user supplied a custom version, use it here. Otherwise, default to crate version.
+    let version_tokens = match metadata.get_custom_version() {
+        Some(s) => {
+            let lit = LitStr::new(s.as_str(), s.as_str().span());
+            quote! {
+                #lit
+            }
+        }
+        None => quote! {
+            env!("CARGO_PKG_VERSION")
+        },
+    };
+
+    quote! {
+        static #static_name_ident: &'static str = #stylized_name_literal;
+        static #static_version_ident: &'static str = #version_tokens;
+    }
+}
+
+/// Takes a capabilities flag and builds the evmc_get_capabilities callback.
+fn build_capabilities_fn(capabilities: u32) -> proc_macro2::TokenStream {
+    let capabilities_literal =
+        LitInt::new(capabilities as u64, IntSuffix::U32, capabilities.span());
```

Comment:
Clippy gives this:
```
warning: casting u32 to u64 may become silently lossy if types change
   --> bindings/rust/evmc-declare/src/lib.rs:281:21
    |
281 |         LitInt::new(capabilities as u64, IntSuffix::U32, capabilities.span());
    |                     ^^^^^^^^^^^^^^^^^^^ help: try: `u64::from(capabilities)`
```

Reply:
This is expected. This is a problem with `syn` because it takes a u64 for any integer literal type. In our case, there should be no issue.

Reply:
Is `u64::from` nicer?